### PR TITLE
Custom user profile

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**

### CODE REVIEW

File: books/views.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    template_name = 'books/index.html'
+    context_object_name = 'latest_book_list'
+
+    def get_queryset(self):
+        """
+        Return the last five published questions (not including those set to be
+        published in the future).
+        """
+
+        return Book.objects.order_by('-id')[:5]
```
Comment (asamolion) [inline books/views.py]: Remove it here.

File: books/views.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class DetailView(generic.DetailView):
+    model = Book
+    template_name = 'books/detail.html'
+
+    def get_queryset(self):
```
Comment (asamolion) [inline books/views.py]: Why is this necessary?

File: userauth/urls.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from django.urls import path
+
+from . import views
+
+app_name = 'userauth'
+urlpatterns = [
+    path('', views.IndexView.as_view(), name='index'),
+    path('login/', views.login, name='login'),
```
Comment (asamolion) [inline userauth/urls.py]: These are unnecessary here. Please move them to the main urls.py file.

File: userauth/views.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+from django.contrib.auth.models import User
+from django.shortcuts import render, redirect
+from django.views import generic
+
+from books.models import Book
+
+
+class IndexView(generic.ListView):
```
Comment (asamolion) [inline userauth/views.py]: This view is unnecessary here.

File: userauth/views.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+        return Book.objects.order_by('-id')[:5]
+
+
+def login(request):
+    if request.POST['username'] == '' or request.POST['password'] == '':
+        return render(request, 'userauth/index.html', {
+            'error_message': "You didn't enter text in a field.",
+        })
+    if User.objects.filter(username__iexact=request.POST['username'], password=request.POST['password']).exists():
```
Comment (asamolion) [inline userauth/views.py]: Do not store the password in plain text.

File: userauth/views.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    else:
+        return render(request, 'userauth/index.html', {
+            'error_message': "USERNAME/PASSWORD MISMATCH.",
+        })
+
+
+def signup(request):
+    if request.POST['username'] == '' or request.POST['password'] == '' or request.POST['email'] == '':
+        return render(request, 'userauth/index.html', {
+            'error_message': "You didn't enter text in a field.",
```
Comment (asamolion) [inline userauth/views.py]: Which field?

File: books/views.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply books/views.py]: it was unnecessary hence it's been removed

File: userauth/views.py (hunk 4 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: it has been replaced with a method

File: userauth/views.py (hunk 5 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: issue is resolved. The password is now encrypted.

File: userauth/views.py (hunk 6 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: Given a proper error message.

File: userauth/forms.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+class UserCreateForm(forms.ModelForm):
+    class Meta:
+        model = User
+        fields = ["username", "password", "email"]
+
+
+class UserLoginForm(forms.Form):
+    username = forms.CharField(widget=forms.TextInput(attrs={'placeholder': 'Username'}))
+    password = forms.CharField(widget=forms.PasswordInput(attrs={'placeholder': 'Password'}))
```
Comment (asamolion) [inline userauth/forms.py]: https://stackoverflow.com/questions/17523263/how-to-create-password-field-in-model-django

File: userauth/urls.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path
+
+from . import views
+
+app_name = 'userauth'
+urlpatterns = [
+    path('', views.index, name='index'),
+    path('signup/', views.signup, name='signup'),
```
Comment (asamolion) [inline userauth/urls.py]: These paths should start from root i.e `/` not `/login/`

File: userauth/views.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+from django.contrib.auth.models import User
+from django.shortcuts import render, redirect
+
+from userauth.forms import UserCreateForm, UserProfileForm, UserLoginForm
+
+
+def index(request):
+    error_message = ''
+    try:
+        request.session['username']
```
Comment (asamolion) [inline userauth/views.py]: There is a method to check whether user is logged in or not. It is provided by django. use it.

File: userauth/views.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    return render(request, 'userauth/index.html', {'user_form': user_form, 'error_message': error_message})
+
+
+def signup(request):
+    user_form = UserCreateForm(data=request.POST)
+    profile_form = UserProfileForm(data=request.POST)
+    if user_form.is_valid() and profile_form.is_valid():
+        user = user_form.save()
+        # form brings back a plain text string, not an encrypted password
+        pw = user.password
```
Comment (asamolion) [inline userauth/views.py]: use more descriptive variable names.

File: userauth/views.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        # thus we need to use set password to encrypt the password string
```
Comment (asamolion) [inline userauth/views.py]: We know this. These comments aren't necessaary.

File: userauth/views.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        user.set_password(pw)
+        user.save()
+        request.session['username'] = user.username
+        login(request, user)
+        return redirect('books:index')
+    return render(request, 'userauth/signup.html', {'user_form': user_form, 'profile_form': profile_form})
+
+
+def logout_user(request):
+    del request.session['username']
```
Comment (asamolion) [inline userauth/views.py]: Is this the official way to logout a user from django?

File: userauth/models.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
+from django.contrib.auth.models import User
+from django.db import models
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+
+class Profile(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    GENDER_CHOICES = [(1, 'Male'), (0, 'Female')]
+    gender = models.IntegerField(choices=GENDER_CHOICES, default=1)
```
Comment (asamolion) [inline userauth/models.py]: Why would you save genders as integers instead of strings?

File: books/views.py (hunk 1 above)
Comment (Ali-D-Akbar) [reply books/views.py]: I'm sorry. it has been handled.

File: userauth/urls.py (hunk 8 above)
Comment (Ali-D-Akbar) [reply userauth/urls.py]: done

File: userauth/views.py (hunk 10 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: done

File: userauth/views.py (hunk 9 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: it can be done by if user.is_authenticated. It's been replaced.

File: userauth/views.py (hunk 11 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: removed unnecessary comments

File: userauth/views.py (hunk 12 above)
Comment (Ali-D-Akbar) [reply userauth/views.py]: it can be done by if user.is_authenticated. It's been replaced.

File: userauth/forms.py (hunk 7 above)
Comment (Ali-D-Akbar) [reply userauth/forms.py]: password type field added

File: userauth/urls.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply userauth/urls.py]: it's been updated in the recent version.

File: userauth/models.py (hunk 13 above)
Comment (Ali-D-Akbar) [reply userauth/models.py]: I'm now using characters 'M' and 'F' to save gender data now

File: userauth/urls.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply userauth/urls.py]: it's been resolved.

File: userauth/views.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    profile_form = UserProfileForm(data=request.POST)
+    if user_form.is_valid() and profile_form.is_valid():
+        user = user_form.save()
+        password = user.password
+        user.set_password(password)
+        user.save()
+        profile = profile_form.save(commit=False)
+        profile.user = user
+        profile.save()
+        request.session['username'] = user.username
```
Comment (asamolion) [inline userauth/views.py]: Why is this necessary?

Comment (Ali-D-Akbar) [reply userauth/views.py]: removed
